src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1377,20 +1368,23 @@ impl Store {
     perms: Permissions,
     is_executable: bool,
     can_be_immutable: bool,
-    immutable_inputs: Option<&ImmutableInputs>,
   ) -> Result<(), StoreError> {
-    if can_be_immutable && digest.size_bytes > IMMUTABLE_FILE_SIZE_LIMIT {
-      let dest_path = immutable_inputs
-        .unwrap()
-        .path_for_file(digest, is_executable)
-        .await?;
-      self
-        .materialize_hardlink(destination, dest_path.to_str().unwrap().to_string())
-        .await
+    let hardlink_dest = if can_be_immutable {
+      self.local.load_from_fs(digest).await?
     } else {
-      self
-        .materialize_file(destination, digest, perms, is_executable)
-        .await
+      None
+    };
+    match hardlink_dest {
+      Some(path) => {
+        self
+          .materialize_hardlink(destination, path.to_str().unwrap().to_string())
+          .await
+      }
```

Comment:
The naming is confusing in here: `hardlink_dest` ends up being the "target" argument to `materialize_hardlink`.